File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 첫번째줄에 격자크기와 바이스러스의 수가 주어진다
// 두번째줄부터 격자가 주어진다
// 마지막줄에 시간제한과 좌표x,y가 주어진다
// 바이러스 숫자가 작은순서대로 상하좌우 전염시킨다
// 그렇다면 큐에 먼저 [virus,x,y,0]를 넣고 sort를 시켜야한다
// 그다음에 시간이 맥스보다 커지면 break를 해줘야한다
// targetX-1, tagetY-1
const solution = (input) => {
  const [n, virusN] = input[0].split(" ").map((v) => Number(v));
  const grid = [];
  const [maxTime, targetX, targetY] = input[n + 1]
    .split(" ")
    .map((v) => Number(v));

  for (let i = 1; i <= n; i++) {
    grid.push(input[i].split(" ").map((v) => Number(v)));
  }

  const queue = [];

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] > 0) {
        queue.push([grid[i][j], i, j, 0]);
      }
    }
  }

  queue.sort((a, b) => a[0] - b[0]);
  const dirs = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  while (queue.length) {
    const [virusNum, x, y, time] = queue.shift();
    if (time === maxTime) break;

    for (const [dx, dy] of dirs) {
      const nextX = x + dx;
      const nextY = y + dy;
      const isValid =
        nextX >= 0 &&
        nextX < n &&
        nextY >= 0 &&
        nextY < n &&
        grid[nextX][nextY] === 0;

      if (isValid) {
        queue.push([virusNum, nextX, nextY, time + 1]);
        grid[nextX][nextY] = virusNum;
      }
    }
  }

  console.log(grid[targetX - 1][targetY - 1]);
};

solution(input);

File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.py
from collections import deque

sizeN, maxNum = list(map(int, input().split()))
grid = []

for i in range(sizeN):
  grid.append(list(map(int, input().split())))

targetTime, targetX, targetY = list(map(int, input().split()))

arr = []
queue =deque()

for i in range(sizeN):
  for j in range(sizeN):
    if grid[i][j] != 0:
      arr.append([i,j,grid[i][j]])

arr.sort(key=lambda x: x[2])

for x,y,val in arr:
  queue.append([x,y,val,0])

dirs = [[-1, 0],[1, 0],[0, -1],[0, 1]]

while len(queue):
  curX, curY, virusNum, time = queue.popleft()
  if time >= targetTime:
    break
  
  for dx, dy in dirs:
    nextX = curX + dx
    nextY = curY + dy
    isValid = 0<=nextX<sizeN and 0<=nextY<sizeN and grid[nextX][nextY] == 0
    
    if isValid:
      queue.append([nextX, nextY, virusNum, time + 1])
      grid[nextX][nextY] = virusNum


print(grid[targetX-1][targetY-1] if grid[targetX-1][targetY-1] !=0 else 0)